### PR TITLE
Cosmos: add partition key to the key

### LINKED ISSUE #20509: Cosmos: add partition key to the PK
Currently it's only added to the alternate key that includes `id`.

Throw if it's not part of any key and issue a warning when using Find when the partition key is not part of the PK

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Add model building tests for Cosmos and fix found bugs.

Fixes #20509

### CODE REVIEW

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs [hunk 1]
```diff
@@ -37,7 +37,7 @@ public DocumentSource([NotNull] IEntityType entityType, [NotNull] CosmosDatabase
         {
             _collectionId = entityType.GetContainer();
             _database = database;
-            _idProperty = entityType.FindProperty(StoreKeyConvention.IdPropertyName);
+            _idProperty = entityType.GetProperties().FirstOrDefault(p => p.GetJsonPropertyName() == StoreKeyConvention.IdPropertyName);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Update/Internal/DocumentSource.cs:40]: What is the difference with FindProperty?

File: src/EFCore/Metadata/Builders/DiscriminatorBuilder.cs [hunk 2]
```diff
@@ -58,7 +58,7 @@ private DiscriminatorBuilder IsComplete(bool complete, ConfigurationSource confi
                     return null;
                 }
 
-                EntityTypeBuilder.Metadata.SetDiscriminatorValue(complete, configurationSource == ConfigurationSource.DataAnnotation);
+                EntityTypeBuilder.Metadata.SetDiscriminatorMappingComplete(complete, configurationSource == ConfigurationSource.DataAnnotation);
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/DiscriminatorBuilder.cs:61]: 😮

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs [hunk 3]
```diff
@@ -57,7 +55,11 @@ public KeyDiscoveryConvention([NotNull] ProviderConventionSetBuilderDependencies
         /// </summary>
         protected virtual ProviderConventionSetBuilderDependencies Dependencies { get; }
 
-        private void TryConfigurePrimaryKey(IConventionEntityTypeBuilder entityTypeBuilder)
+        /// <summary>
+        ///     Discovers primary key candidates and configures the primary key if found
```
Comment (smitpatel) [inline src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:59]:
```suggestion
        ///     Discovers primary key candidates and configures the primary key if found.
```

File: src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs [hunk 4]
```diff
@@ -99,6 +101,7 @@ private void TryConfigurePrimaryKey(IConventionEntityTypeBuilder entityTypeBuild
                 if (keyProperties.Count == 0
                     || definingFk.Properties.Contains(keyProperties.First()))
                 {
+                    var primaryKey = entityType.FindPrimaryKey();
```
Comment (smitpatel) [inline src/EFCore/Metadata/Conventions/KeyDiscoveryConvention.cs:104]: Either use variable or remove it.

File: test/EFCore.Cosmos.Tests/ModelBuilding/CosmosModelBuilderGenericTest.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                var entity = model.FindEntityType(typeof(Customer));
+
+                Assert.Equal(new[] { nameof(Customer.Id), nameof(Customer.AlternateKey) },
+                    entity.FindPrimaryKey().Properties.Select(p => p.Name));
+                Assert.Equal(new[] { "id", nameof(Customer.AlternateKey) },
+                    entity.GetKeys().First(k => k != entity.FindPrimaryKey()).Properties.Select(p => p.Name));
+            }
+
+            [ConditionalFact]
+            public virtual void Partition_key_is_added_to_the_alternate_key()
```
Comment (smitpatel) [inline test/EFCore.Cosmos.Tests/ModelBuilding/CosmosModelBuilderGenericTest.cs:53]: What is different from previous test?

File: test/EFCore.Cosmos.Tests/ModelBuilding/CosmosModelBuilderGenericTest.cs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        public class CosmosGenericInheritance : GenericInheritance
+        {
+            public override void Can_set_and_remove_base_type()
+            {
+                // Fails due to presence of __jObject
+            }
+
+            public class Parent
```
Comment (smitpatel) [inline test/EFCore.Cosmos.Tests/ModelBuilding/CosmosModelBuilderGenericTest.cs:146]: Where are these classes used?

File: test/EFCore.Tests/ModelBuilding/OneToManyTestBase.cs [hunk 7]
```diff
@@ -389,21 +346,7 @@ public virtual void Creates_relationship_with_no_navigations()
                 Assert.Equal(fk.PrincipalKey.Properties, newFk.PrincipalKey.Properties);
                 Assert.Empty(dependentType.GetNavigations().Where(nav => nav.ForeignKey == newFk));
                 Assert.Empty(principalType.GetNavigations().Where(nav => nav.ForeignKey == newFk));
-                AssertEqual(
```
Comment (smitpatel) [inline test/EFCore.Tests/ModelBuilding/OneToManyTestBase.cs:392]: I think in the other class you added Assert.Contains for this.

File: src/EFCore.Cosmos/Update/Internal/DocumentSource.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Update/Internal/DocumentSource.cs:40]: One might have a "MyEntityId" mapped to "id"

**Pull request merged**